core/utils/utils_test.go

```diff
@@ -469,3 +469,74 @@ func Test_WithJitter(t *testing.T) {
 		require.LessOrEqual(t, int(r), int(11*time.Second))
 	}
 }
+
+func Test_StartStopOnce_StopWaitsForStartToFinish(t *testing.T) {
+	t.Parallel()
+
+	once := utils.StartStopOnce{}
+
+	ch := make(chan int, 3)
+
+	ready := make(chan bool)
+
+	go func() {
+		once.StartOnce("slow service", func() (err error) {
+			ch <- 1
+			ready <- true
+			<-time.After(2 * time.Second)
+			ch <- 2
+
+			return nil
+		})
+
+	}()
+
+	go func() {
+		<-ready // try stopping halfway through startup
+		once.StopOnce("slow service", func() (err error) {
+			ch <- 3
+
+			return nil
+		})
+
+	}()
+
+	require.Equal(t, 1, <-ch)
+	require.Equal(t, 2, <-ch)
+	require.Equal(t, 3, <-ch)
+}
+
+func Test_StartStopOnce_MultipleStartNoBlock(t *testing.T) {
+	t.Parallel()
+
+	once := utils.StartStopOnce{}
+
+	ch := make(chan int, 3)
+
+	ready := make(chan bool)
+
+	go func() {
+		ch <- 1
+		once.StartOnce("slow service", func() (err error) {
+			ready <- true
+			<-time.After(2 * time.Second)
```

Comment:
Is there any way we can avoid a sleep of 2 seconds? Perhaps by blocking a channel in the caller goroutine?

Reply:
I've removed it on this test, and lowered the delay on the other one (can't emit a signal inside/after StopOnce there since the call will block until StartOnce completes)